File: frontend/src/app/core/auth/auth.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable, computed, signal } from '@angular/core';
import { Observable, map, tap } from 'rxjs';

import { environment } from '../../../environments/environment';
import { LoginRequest, RegistroRequest, TokenResponse, UsuarioOut } from './auth.models';

const STORAGE_KEY = 'fashionstore.sesion';

interface SesionGuardada {
  token: string;
  usuario: UsuarioOut;
}

@Injectable({ providedIn: 'root' })
export class AuthService {
  private readonly sesion = signal<SesionGuardada | null>(this.leerSesionGuardada());

  readonly usuario = computed(() => this.sesion()?.usuario ?? null);
  readonly estaAutenticado = computed(() => this.sesion() !== null);
  readonly esStaff = computed(() => this.usuario()?.tipo === 'STAFF');
  readonly permisos = computed(() => this.usuario()?.permisos ?? []);

  constructor(private readonly http: HttpClient) {}

  get token(): string | null {
    return this.sesion()?.token ?? null;
  }

  /** CU13: la UI se arma con los permisos del rol, no con su nombre. Basta con tener uno. */
  tienePermiso(...codigos: string[]): boolean {
    const propios = this.permisos();
    return codigos.some((codigo) => propios.includes(codigo));
  }

  iniciarSesion(credenciales: LoginRequest): Observable<UsuarioOut> {
    return this.http.post<TokenResponse>(`${environment.apiUrl}/auth/login`, credenciales).pipe(
      tap((respuesta) => this.guardar(respuesta)),
      map((respuesta) => respuesta.usuario),
    );
  }

  registrarse(datos: RegistroRequest): Observable<UsuarioOut> {
    return this.http.post<TokenResponse>(`${environment.apiUrl}/auth/registro`, datos).pipe(
      tap((respuesta) => this.guardar(respuesta)),
      map((respuesta) => respuesta.usuario),
    );
  }

  /**
   * Vuelve a pedir el usuario al backend. Si el administrador cambio el rol o los permisos
   * (CU13), la sesion guardada en el navegador queda desactualizada hasta este refresco.
   */
  refrescarSesion(): Observable<UsuarioOut> {
    return this.http.get<UsuarioOut>(`${environment.apiUrl}/auth/yo`).pipe(
      tap((usuario) => {
        const actual = this.sesion();
        if (!actual) return;
        const guardada: SesionGuardada = { token: actual.token, usuario };
        this.sesion.set(guardada);
        localStorage.setItem(STORAGE_KEY, JSON.stringify(guardada));
      }),
    );
  }

  cerrarSesion(): void {
    this.sesion.set(null);
    localStorage.removeItem(STORAGE_KEY);
  }

  private guardar(respuesta: TokenResponse): void {
    const guardada: SesionGuardada = { token: respuesta.access_token, usuario: respuesta.usuario };
    this.sesion.set(guardada);
    localStorage.setItem(STORAGE_KEY, JSON.stringify(guardada));
  }

  private leerSesionGuardada(): SesionGuardada | null {
    const crudo = localStorage.getItem(STORAGE_KEY);
    if (!crudo) return null;
    try {
      const guardada = JSON.parse(crudo) as SesionGuardada;
      // sesiones guardadas antes de CU13 no traen permisos
      guardada.usuario.permisos ??= [];
      return guardada;
    } catch {
      localStorage.removeItem(STORAGE_KEY);
      return null;
    }
  }
}

File: movil/lib/core/auth/auth_service.dart
import 'dart:convert';

import 'package:flutter/foundation.dart';
import 'package:shared_preferences/shared_preferences.dart';

import '../api.dart';
import 'auth_models.dart';

/// Sesion del usuario. Espejo de `AuthService` del frontend: guarda token + usuario en
/// el almacenamiento local (aca SharedPreferences en vez de localStorage) y lo restaura
/// al arrancar la app.
class AuthService extends ChangeNotifier {
  static const String _claveSesion = 'fashionstore.sesion';

  String? _token;
  UsuarioOut? _usuario;
  bool _listo = false;

  String? get token => _token;
  UsuarioOut? get usuario => _usuario;
  bool get estaAutenticado => _token != null && _usuario != null;
  bool get esStaff => _usuario?.esStaff ?? false;
  List<String> get permisos => _usuario?.permisos ?? const [];

  /// Falso hasta que se termino de leer la sesion guardada (evita parpadeos al iniciar).
  bool get listo => _listo;

  /// CU13: la UI se arma con los permisos del rol, no con su nombre. Basta con tener uno.
  bool tienePermiso(List<String> codigos) => codigos.any(permisos.contains);

  Future<void> inicializar() async {
    api.obtenerToken = () => _token;
    api.alExpirarSesion = () {
      if (estaAutenticado) cerrarSesion();
    };

    final prefs = await SharedPreferences.getInstance();
    final crudo = prefs.getString(_claveSesion);
    if (crudo != null) {
      try {
        final datos = jsonDecode(crudo) as Map<String, dynamic>;
        _token = datos['token'] as String?;
        _usuario = UsuarioOut.desdeJson(datos['usuario'] as Map<String, dynamic>);
      } catch (_) {
        await prefs.remove(_claveSesion);
        _token = null;
        _usuario = null;
      }
    }
    _listo = true;
    notifyListeners();
  }

  Future<UsuarioOut> iniciarSesion(String email, String password) async {
    final respuesta = await api.post('/auth/login', cuerpo: {'email': email, 'password': password});
    return _guardar(respuesta as Map<String, dynamic>);
  }

  Future<UsuarioOut> registrarse(RegistroRequest datos) async {
    final respuesta = await api.post('/auth/registro', cuerpo: datos.aJson());
    return _guardar(respuesta as Map<String, dynamic>);
  }

  /// Vuelve a pedir el usuario al backend: si el administrador cambio su rol o sus
  /// permisos (CU13), la sesion guardada en el dispositivo quedo desactualizada.
  Future<void> refrescarSesion() async {
    if (!estaAutenticado) return;
    final respuesta = await api.get('/auth/yo');
    _usuario = UsuarioOut.desdeJson(respuesta as Map<String, dynamic>);
    await _persistir();
    notifyListeners();
  }

  Future<void> cerrarSesion() async {
    _token = null;
    _usuario = null;
    notifyListeners();
    final prefs = await SharedPreferences.getInstance();
    await prefs.remove(_claveSesion);
  }

  Future<UsuarioOut> _guardar(Map<String, dynamic> respuesta) async {
    _token = respuesta['access_token'] as String;
    _usuario = UsuarioOut.desdeJson(respuesta['usuario'] as Map<String, dynamic>);
    await _persistir();
    notifyListeners();
    return _usuario!;
  }

  Future<void> _persistir() async {
    if (_token == null || _usuario == null) return;
    final prefs = await SharedPreferences.getInstance();
    await prefs.setString(
      _claveSesion,
      jsonEncode({'token': _token, 'usuario': _usuario!.aJson()}),
    );
  }
}
